sql/optbuilder/doc.go

```diff
@@ -0,0 +1,148 @@
+// those in the GROUP BY and SELECT target lists.
+//
+// In the example below, we identify and tag two aggregations that are assigned
+// expression ids after x,y, and z: SUM(y):4, COUNT(y):5. The sort node and
+// target list projections that use those aggregations resolve dependencies by
+// id reference:
+//
+// select x, sum(y) from xy group by x order by x - count(y)
+// =>
+// Project
+// ├─ columns: [xy.x:1!null, SUM(xy.y):4!null as sum(y)]
+// └─ Sort((xy.x:1!null - COUNT(xy.y):5!null) ASC nullsFirst)
+//     └─ GroupBy
+//         ├─ select: xy.y:2!null, xy.x:1!null, SUM(xy.y:2!null), COUNT(xy.y:2!null)
+//         ├─ group: xy.x:1!null
+//         └─ Table
+//             ├─ name: xy
+//             └─ columns: [x y z]
+//
+// Passthrough columns not included in the SELECT target list are included in the
+// aggregation outputs:
+//
+// select x from xy having z > 0
+//
+// Project
+// ├─ columns: [xy.x:1!null]
+// └─ Having
+//     ├─ GreaterThan
+//     │   ├─ xy.z:3!null
+//     │   └─ 0 (tinyint)
+//     └─ GroupBy
+//         ├─ select: xy.x:1!null, xy.z:3!null
+//         ├─ group:
+//         └─ Table
+//             ├─ name: xy
+//             └─ columns: [x y z]
+//
+// Aggregations are probably a long-tail of testing to get this behavior right,
+// particularly when aggregate functions are initialized outside of their
+// execution scope (select (select u from uv where max(x) > u limit 1) from xy).
+//
+// The second difficulty is how to represent expressions and references while
+// building the plan, and how low in the tree to execute expression logic. This
+// is a secondary concern compared to generating unique ids for aggregation
+// functions and source columns.
+//
+// For example, (x+z) is a target and grouping column below. The aggregation
+// could return (x+z) which the target list passes through:
+//
+// SELECT count(xy.x) AS count_1, x + z AS lx FROM xy GROUP BY x + z
+// =>
+// Project
+// ├─ columns: [COUNT(xy.x):4!null as count_1, (xy.x:1!null + xy.z:3!null) as lx]
+// └─ GroupBy
+//     ├─ select: xy.x:1!null, (xy.x:1!null + xy.z:3!null), COUNT(xy.x:1!null), xy.z:3!null
+//     ├─ group: (xy.x:1!null + xy.z:3!null)
+//     └─ Table
+//         ├─ name: xy
+//         └─ columns: [x y z]
+//
+// We do not have a good way of referencing expressions that are not
+// aggregation functions or table columns. In other databases, expressions are
+// interned when they are added to the memo. So an expression will be evaluated
+// and available for reference at the lowest level of the tree it was built. If
+// an aggregation builds an expression, a projection built later will find
+// the reference and avoid re-computing the value. If a relation earlier in the
+// tree built a subtree of an expression currently being built, it can input
+// the reference rather than computing the subtree.
+//
+// Questions:
+// - can all node types be resolved with this setup? do we run into problems
```

Comment:
Single nesting subqueries and sq expressions work pretty well right now. A scope has a reference to the parent scope, so we fall upwards looking for definitions. The resolveColumn/resolveOuter distinction is weird right now, it was a hack because there were some cases where we expect only a same-scope definition. I still have to do plan-wide expression ids to make subqueries more sensible. Also I might have to inline cacheability up-front. That would sort of look like what I did with aggregations, where I add an attribute like `outerColumns` during analyze that populates `cacheable` in build.